firmware/src/imu/mod.rs

```diff
@@ -12,13 +12,44 @@ use crate::{
 };
 
 pub type Quat = nalgebra::UnitQuaternion<f32>;
+pub type Accel = nalgebra::Vector3<f32>;
+pub type Gyro = nalgebra::Vector3<f32>;
 
-pub trait FusedImu {
-	type Error: core::fmt::Debug;
+pub struct UnfusedData {
+	accel: Accel,
+	gyro: Gyro,
+}
+
+pub struct FusedData {
+	quat: Quat,
+}
+
+/// Represents a sensor fusion algorithm that will take an imu's `UnfusedData` and do math to turn
+/// it into `FusedData`, suitable for use as orientation.
+pub trait Fuser {
+	// TODO: Does a one-in, one-out api here work? Should we reuse a standard trait like a
+	// sink/stream/iterator?
+	// Note: Intentionally not async rn, this should only be doing math, not io or any internal
+	// awaiting.
+	fn process(&mut self, unfused: &UnfusedData) -> FusedData;
+}
+
+pub trait Imu {
+	type Error: core::fmt::Debug; // TODO: Maybe use defmt instead?
+	/// The data that the imu outputs.
+	type Data;
```

Comment:
So there were two main options that I was considering - two separate traits, or one trait with associated types.

If I used two separate traits, i could do it like this:
```rust
trait FusedImu {
  type Error;

  /// Performs IO to get the next data from the imu.
  async fn next_fused(&mut self) -> Result<FusedData, Self::Error>;
}

trait UnfusedImu {
  type Error;
  async fn next_unfused(&mut self) -> Result<UnfusedData, Self::Error>;
}
```

But then, it seemed that there should be a relationship between the two. We don't want a `FusedImu` struct to impl `UnfusedImu` because that would allow the user to call that and therefore drop data. Instead it should only be implementing `FusedImu`.

So now the two traits can't have a relationship. And yet most of their function sigs are the same, the only differencee is that they output a different data type.

Instead, having an associated type for that data seems simpler. We can always introduce a trait alias if we really want to, using the same technique we did before for the other trait aliases. So we can then specify a `FusedImu` instead of a `Imu<Data=FusedData>` if we are really worried about that trait being confusing.

But I think its simple enough and avoid the inherent complexity in what would otherwise be basically a trait inheritance issue.